src/Abp/EntityHistory/EntitySnapshotManagerBase.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+using Abp.Dependency;
+using Abp.Domain.Entities;
+using Abp.Json;
+
+namespace Abp.EntityHistory
+{
+    public abstract class EntitySnapshotManagerBase : IEntitySnapshotManager, ITransientDependency
+    {
+        protected abstract Task<TEntity> GetEntityById<TEntity, TPrimaryKey>(TPrimaryKey id)
+            where TEntity : class, IEntity<TPrimaryKey>;
+
+        protected abstract IQueryable<EntityChange> GetEntityChanges<TEntity, TPrimaryKey>(TPrimaryKey id,
+            DateTime snapshotTime) where TEntity : class, IEntity<TPrimaryKey>;
+
+        public async Task<EntityHistorySnapshot> GetSnapshotAsync<TEntity>(int id, DateTime snapshotTime) where TEntity : class, IEntity<int>
+        {
+            return await GetSnapshotAsync<TEntity, int>(id, snapshotTime);
+        }
+        public virtual async Task<EntityHistorySnapshot> GetSnapshotAsync<TEntity, TPrimaryKey>(TPrimaryKey id, DateTime snapshotTime) where TEntity : class, IEntity<TPrimaryKey>
+        {
+            var entity = await GetEntityById<TEntity, TPrimaryKey>(id);
+
+            var snapshotPropertiesDictionary = new Dictionary<string, string>();
+
+            var propertyChangesStackTreeDictionary = new Dictionary<string, string>();
+
+            if (entity != null)
+            {
+                var changes = GetEntityChanges<TEntity, TPrimaryKey>(id, snapshotTime)
```

Comment:
probably it is good to use `IAsyncQueryableExecuter` here since this method suppose to be async.

```csharp
// class property
public IAsyncQueryableExecuter AsyncQueryableExecuter { get; set; }

// within the method
var query = GetEntityChanges<TEntity, TPrimaryKey>(id, snapshotTime)..Select(x => new { x.ChangeType, x.PropertyChanges });
var changes = await AsyncQueryableExecuter.ToListAsync(query);
```

Reply:
That's what I was looking for 👍